daemon/cluster/cluster.go

```diff
@@ -979,7 +1021,20 @@ func (c *Cluster) UpdateService(serviceIDOrName string, version uint64, spec typ
 		if ctnr == nil {
 			return fmt.Errorf("service does not use container tasks")
 		}
-		serviceSpec.Task.GetContainer().PullOptions = ctnr.PullOptions
+		newCtnr.PullOptions = ctnr.PullOptions
+		// update encodedAuth so it can be used to pin image by digest
+		if ctnr.PullOptions != nil {
+			encodedAuth = ctnr.PullOptions.RegistryAuth
+		}
+	}
+
+	// pin image by digest
+	digestImage, err := c.imageWithDigestString(ctx, newCtnr.Image, encodedAuth)
+	if err == nil {
+		logrus.Debugf("pinning image %s by digest: %s", newCtnr.Image, digestImage)
+		newCtnr.Image = digestImage
+	} else {
+		logrus.Debugf("unable to pin image %s to digest: %s", newCtnr.Image, err.Error())
```

Comment:
This would log an error message if the user already provided a digest. How about making `imageWithDigestString` return `image, nil` if the image already has a digest specified, and then this code can only log "pinning image..." if `digestImage != newCntr.Image`?